
# **Análisis de la variación anual de la temperatura en superficie**

Este indicador presenta el cambio medio de la temperatura en superficie durante el periodo 1961-2022, utilizando como referencia las temperaturas entre 1951 y 1980. Utilice los menús desplegables para buscar los cambios de temperatura por país.

Estos datos proceden de la Base de Datos Estadísticos Corporativos de la Organización de las Naciones Unidas para la Agricultura y la Alimentación (FAOSTAT) y se basan en datos GISTEMP de acceso público del Instituto Goddard de Estudios Espaciales de la Administración Nacional de Aeronáutica y del Espacio (NASA GISS).

Se presentan estimaciones de los cambios en la temperatura media de la superficie, en grados centígrados, para los años 1961-2022 por países y para el mundo.

**Fuente:** Food and Agriculture Organization of the United Nations (FAO). 2022. FAOSTAT Climate Change, Climate Indicators, Temperature change. License: CC BY-NC-SA 3.0 IGO. Extracted from: https://www.fao.org/faostat/en/#data/ET. Accessed on 2023-03-28.

https://climatedata.imf.org/pages/climatechange-data

https://climatedata.imf.org/datasets/4063314923d74187be9596f10d034914/explore


# **1. Exploración del dataset de datos**

In [120]:
import pandas as pd

In [121]:
df = pd.read_csv("Annual_Surface_Temperature_Change.csv")
df.head()

ObjectId                       Country ISO2 ISO3  \
0         1  Afghanistan, Islamic Rep. of   AF  AFG   
1         2                       Albania   AL  ALB   
2         3                       Algeria   DZ  DZA   
3         4                American Samoa   AS  ASM   
4         5      Andorra, Principality of   AD  AND   

                                           Indicator            Unit  \
0  Temperature change with respect to a baseline ...  Degree Celsius   
1  Temperature change with respect to a baseline ...  Degree Celsius   
2  Temperature change with respect to a baseline ...  Degree Celsius   
3  Temperature change with respect to a baseline ...  Degree Celsius   
4  Temperature change with respect to a baseline ...  Degree Celsius   

                                              Source CTS_Code  \
0  Food and Agriculture Organization of the Unite...     ECCS   
1  Food and Agriculture Organization of the Unite...     ECCS   
2  Food and Agriculture Organization of the Unite...     ECCS   
3  Food and Agriculture Organization of the Unite...     ECCS   
4  Food and Agriculture Organization of the Unite...     ECCS   

                     CTS_Name  \
0  Surface Temperature Change   
1  Surface Temperature Change   
2  Surface Temperature Change   
3  Surface Temperature Change   
4  Surface Temperature Change   

                                 CTS_Full_Descriptor  ...  F2013  F2014  \
0  Environment, Climate Change, Climate Indicator...  ...  1.281  0.456   
1  Environment, Climate Change, Climate Indicator...  ...  1.333  1.198   
2  Environment, Climate Change, Climate Indicator...  ...  1.192  1.690   
3  Environment, Climate Change, Climate Indicator...  ...  1.257  1.170   
4  Environment, Climate Change, Climate Indicator...  ...  0.831  1.946   

   F2015  F2016  F2017  F2018  F2019  F2020  F2021  F2022  
0  1.093  1.555  1.540  1.544  0.910  0.498  1.327  2.012  
1  1.569  1.464  1.121  2.028  1.675  1.498  1.536  1.518  
2  1.121  1.757  1.512  1.210  1.115  1.926  2.330  1.688  
3  1.009  1.539  1.435  1.189  1.539  1.430  1.268  1.256  
4  1.690  1.990  1.925  1.919  1.964  2.562  1.533  3.243  

[5 rows x 72 columns]

In [122]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 225 entries, 0 to 224
Data columns (total 72 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ObjectId             225 non-null    int64  
 1   Country              225 non-null    object 
 2   ISO2                 223 non-null    object 
 3   ISO3                 225 non-null    object 
 4   Indicator            225 non-null    object 
 5   Unit                 225 non-null    object 
 6   Source               225 non-null    object 
 7   CTS_Code             225 non-null    object 
 8   CTS_Name             225 non-null    object 
 9   CTS_Full_Descriptor  225 non-null    object 
 10  F1961                188 non-null    float64
 11  F1962                189 non-null    float64
 12  F1963                188 non-null    float64
 13  F1964                188 non-null    float64
 14  F1965                188 non-null    float64
 15  F1966                192 non-null    flo

El conjunto de datos se encuentra compuesto por 72 columnas y 225 registros. El listado de atributos es:
* Country              (225 non-null) [string]
* ISO2                 (223 non-null) [string]
* ISO3                 (225 non-null) [string]
* Indicator            (225 non-null) [string]
* Unit                 (225 non-null) [string]
* Source               (225 non-null) [string]
* CTS_Code             (225 non-null) [string]
* CTS_Name             (225 non-null) [string]
* CTS_Full_Descriptor  (225 non-null) [string]
* Year                 [float64]

El conjunto de datos representa la relación entre la variable numérica temperatura, indicada para cada columna por año, y la variable categórica "Country", país entre los años 1961 y 2022.

Observaciones:
* El alcance de observaciones del dataset es de 225 países.
* Las observaciones para los años no están disponibles para todos los países.
* Las variables categóricas como "Indicator", "CTS_Code", "Unit", "Source", "CTS_Full_Descriptor" únicamente poseen un valor informado.

In [123]:
df.describe()

ObjectId       F1961       F1962       F1963       F1964       F1965  \
count  225.000000  188.000000  189.000000  188.000000  188.000000  188.000000   
mean   113.000000    0.163053   -0.013476   -0.006043   -0.070059   -0.247027   
std     65.096083    0.405080    0.341812    0.387348    0.309305    0.270734   
min      1.000000   -0.694000   -0.908000   -1.270000   -0.877000   -1.064000   
25%     57.000000   -0.097000   -0.164000   -0.205500   -0.236500   -0.392500   
50%    113.000000    0.064500   -0.056000   -0.003000   -0.056000   -0.230500   
75%    169.000000    0.318500    0.114000    0.230500    0.132500   -0.091500   
max    225.000000    1.892000    0.998000    1.202000    1.097000    0.857000   

            F1966       F1967       F1968       F1969  ...       F2013  \
count  192.000000  191.000000  191.000000  190.000000  ...  216.000000   
mean     0.105505   -0.110832   -0.199110    0.157942  ...    0.931199   
std      0.378423    0.339484    0.270131    0.308540  ...    0.321595   
min     -1.801000   -1.048000   -1.634000   -0.900000  ...    0.118000   
25%     -0.035750   -0.259500   -0.340000   -0.009000  ...    0.743500   
50%      0.098000   -0.146000   -0.187000    0.204000  ...    0.897000   
75%      0.277000    0.015000   -0.067000    0.349000  ...    1.187500   
max      1.151000    1.134000    0.476000    0.939000  ...    1.643000   

            F2014       F2015       F2016       F2017       F2018       F2019  \
count  216.000000  216.000000  213.000000  214.000000  213.000000  213.000000   
mean     1.114815    1.269773    1.439521    1.280785    1.302113    1.443061   
std      0.564903    0.462162    0.401091    0.393999    0.596786    0.467510   
min     -0.092000   -0.430000    0.250000    0.017000    0.238000    0.050000   
25%      0.744000    1.017750    1.147000    1.027500    0.865000    1.169000   
50%      0.986500    1.215000    1.446000    1.282000    1.125000    1.412000   
75%      1.335500    1.520500    1.714000    1.535000    1.834000    1.698000   
max      2.704000    2.613000    2.459000    2.493000    2.772000    2.689000   

            F2020       F2021       F2022  
count  212.000000  213.000000  213.000000  
mean     1.552038    1.343531    1.382113  
std      0.621930    0.484692    0.669279  
min      0.229000   -0.425000   -1.305000  
25%      1.161750    1.019000    0.878000  
50%      1.477000    1.327000    1.315000  
75%      1.826250    1.629000    1.918000  
max      3.691000    2.676000    3.243000  

[8 rows x 63 columns]

### Porcentaje de completitud de la información

In [124]:
((len(df) - df.isna().sum()) / len(df)) * 100

ObjectId     100.000000
Country      100.000000
ISO2          99.111111
ISO3         100.000000
Indicator    100.000000
                ...    
F2018         94.666667
F2019         94.666667
F2020         94.222222
F2021         94.666667
F2022         94.666667
Length: 72, dtype: float64

### Comprobación de valores para variables categóricas

In [125]:
df.Indicator.value_counts()

Temperature change with respect to a baseline climatology, corresponding to the period 1951-1980    225
Name: Indicator, dtype: int64

In [126]:
df.CTS_Code.value_counts()

ECCS    225
Name: CTS_Code, dtype: int64

In [127]:
df.Unit.value_counts()

Degree Celsius    225
Name: Unit, dtype: int64

In [128]:
df.Source.value_counts()

Food and Agriculture Organization of the United Nations (FAO). 2022. FAOSTAT Climate Change, Climate Indicators, Temperature change. License: CC BY-NC-SA 3.0 IGO. Extracted from: https://www.fao.org/faostat/en/#data/ET. Accessed on 2023-03-28.    225
Name: Source, dtype: int64

In [129]:
df.CTS_Full_Descriptor.value_counts()

Environment, Climate Change, Climate Indicators, Surface Temperature Change    225
Name: CTS_Full_Descriptor, dtype: int64

### Verificación de integridad de clave primaria

In [130]:
df.Country.isna().sum()

0

# **2. Transformación del dataset**

#### Se seleccionan únicamente las columnas necesarias para nuestro análisis, el país y todas las variables numéricas, es decir el cambio de temperaturan en grados Celsius por cada año.

In [131]:
columns = ['Country']
columns.extend(df.select_dtypes(include=['float64']).columns.to_list())
df_cc = df[columns]

#### Se realiza la despivotización de la tabla. [Ejemplo](https://pandas.pydata.org/docs/dev/_images/reshaping_melt.png)

El DataFrame resultante tendría 13950 registros y las siguientes columnas:
* Country - País de la observación (13950 non-null)  [string]
* Year - Año de la observación (13950 non-null)  [string]
* TemperatureChange - Valor del cambio de temperatura en grados Celsius (12460 non-null)  [float64]


In [132]:
df_result = df_cc.melt(id_vars="Country", var_name="Year", value_name="TemperatureChange")
df_result.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13950 entries, 0 to 13949
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Country            13950 non-null  object 
 1   Year               13950 non-null  object 
 2   TemperatureChange  12460 non-null  float64
dtypes: float64(1), object(2)
memory usage: 327.1+ KB


In [133]:
df_result.head()

Country   Year  TemperatureChange
0  Afghanistan, Islamic Rep. of  F1961             -0.113
1                       Albania  F1961              0.627
2                       Algeria  F1961              0.164
3                American Samoa  F1961              0.079
4      Andorra, Principality of  F1961              0.736

#### Transformaciones a realizar para la preparación final del dataset:
* Eliminar prefijo del valor de la variable año para convertirlo en formato entero.
* Realizar la conversión de nulos a ceros en la variable de cambio de temperatura.

In [134]:
df_result.Year = df_result.Year.str.removeprefix("F").astype('int')

In [135]:
df_result.fillna(0, inplace=True)

In [136]:
df_result.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13950 entries, 0 to 13949
Data columns (total 3 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Country            13950 non-null  object 
 1   Year               13950 non-null  int64  
 2   TemperatureChange  13950 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 327.1+ KB


In [137]:
df_result[df_result.TemperatureChange.isna()]

Empty DataFrame
Columns: [Country, Year, TemperatureChange]
Index: []

In [138]:
df_result.head()

Country  Year  TemperatureChange
0  Afghanistan, Islamic Rep. of  1961             -0.113
1                       Albania  1961              0.627
2                       Algeria  1961              0.164
3                American Samoa  1961              0.079
4      Andorra, Principality of  1961              0.736

In [140]:
import numpy as np
df_result['Size'] = np.where(df_result['TemperatureChange'] <= -2.0, 34,
                             np.where((df_result['TemperatureChange'] > -2.0) & (df_result['TemperatureChange'] <= -1.75), 30,
                                      np.where((df_result['TemperatureChange'] > -1.75) & (df_result['TemperatureChange'] <= -1.5), 26,
                                      np.where((df_result['TemperatureChange'] > -1.5) & (df_result['TemperatureChange'] <= -1.25), 22,
                                               np.where((df_result['TemperatureChange'] > -1.25) & (df_result['TemperatureChange'] <= -1.0), 18,
                                               np.where((df_result['TemperatureChange'] > -1.0) & (df_result['TemperatureChange'] <= -0.75), 14,
                                                        np.where((df_result['TemperatureChange'] > -0.75) & (df_result['TemperatureChange'] <= -0.5), 10,
                                                        np.where((df_result['TemperatureChange'] > -0.5) & (df_result['TemperatureChange'] <= -0.25), 6,
                                                                 np.where((df_result['TemperatureChange'] > -0.25) & (df_result['TemperatureChange'] < 0.0), 2,
                                                                          np.where((df_result['TemperatureChange'] >= 0.0) & (df_result['TemperatureChange'] < 0.25), 2,
                                                                          np.where((df_result['TemperatureChange'] >= 0.25) & (df_result['TemperatureChange'] < 0.75), 6,
                                                                          np.where((df_result['TemperatureChange'] >= 0.75) & (df_result['TemperatureChange'] < 1.0), 10,
                                                                          np.where((df_result['TemperatureChange'] >= 1.0) & (df_result['TemperatureChange'] < 1.25), 14,
                                                                          np.where((df_result['TemperatureChange'] >= 1.25) & (df_result['TemperatureChange'] < 1.5), 18,
                                                                          np.where((df_result['TemperatureChange'] >= 1.5) & (df_result['TemperatureChange'] < 1.75), 22,
                                                                          np.where((df_result['TemperatureChange'] >= 1.75) & (df_result['TemperatureChange'] < 2.0), 26,
                                                                          np.where((df_result['TemperatureChange'] >= 2.0) & (df_result['TemperatureChange'] < 2.25), 30,
                                                                          np.where((df_result['TemperatureChange'] >= 2.25) & (df_result['TemperatureChange'] < 2.5), 34,
                                                                          np.where((df_result['TemperatureChange'] >= 2.5) & (df_result['TemperatureChange'] < 2.75), 38,
                                                                          np.where((df_result['TemperatureChange'] >= 2.75) & (df_result['TemperatureChange'] < 3.0), 42,
                                                                          np.where((df_result['TemperatureChange'] >= 3.0) & (df_result['TemperatureChange'] < 3.25), 46,
                                                                          np.where((df_result['TemperatureChange'] >= 3.25) & (df_result['TemperatureChange'] < 3.5), 50,
                                                                          np.where(df_result['TemperatureChange'] >= 3.5, 54,
                                                                 2)))))))))))))))))))))))
df_result['Color'] = np.where(df_result['TemperatureChange'] <= -2.0, '<-2.0',
                             np.where((df_result['TemperatureChange'] > -2.0) & (df_result['TemperatureChange'] <= -1.75), '-2.0 to -1.75',
                                      np.where((df_result['TemperatureChange'] > -1.75) & (df_result['TemperatureChange'] <= -1.5), '-1.75 to -1.50',
                                      np.where((df_result['TemperatureChange'] > -1.5) & (df_result['TemperatureChange'] <= -1.25), '-1.50 to -1.25',
                                               np.where((df_result['TemperatureChange'] > -1.25) & (df_result['TemperatureChange'] <= -1.0), '-1.25 to -1.0',
                                               np.where((df_result['TemperatureChange'] > -1.0) & (df_result['TemperatureChange'] <= -0.75), '-1.0 to -0.75',
                                                        np.where((df_result['TemperatureChange'] > -0.75) & (df_result['TemperatureChange'] <= -0.5), '-0.75 to -0.25',
                                                        np.where((df_result['TemperatureChange'] > -0.5) & (df_result['TemperatureChange'] <= -0.25), '-0.50 to 0.25',
                                                                 np.where((df_result['TemperatureChange'] > -0.25) & (df_result['TemperatureChange'] < 0.0), '-0.25 to 0.0',
                                                                          np.where((df_result['TemperatureChange'] >= 0.0) & (df_result['TemperatureChange'] < 0.25), '0.0 to 0.25',
                                                                          np.where((df_result['TemperatureChange'] >= 0.25) & (df_result['TemperatureChange'] < 0.5), '0.25 to 0.50',
                                                                          np.where((df_result['TemperatureChange'] >= 0.5) & (df_result['TemperatureChange'] < 0.75), '0.50 to 0.75',
                                                                          np.where((df_result['TemperatureChange'] >= 0.75) & (df_result['TemperatureChange'] < 1.0), '0.75 to 1.0',
                                                                          np.where((df_result['TemperatureChange'] >= 1.0) & (df_result['TemperatureChange'] < 1.25), '1.0 to 1.25',
                                                                          np.where((df_result['TemperatureChange'] >= 1.25) & (df_result['TemperatureChange'] < 1.5), '1.25 to 1.50',
                                                                          np.where((df_result['TemperatureChange'] >= 1.5) & (df_result['TemperatureChange'] < 1.75), '1.50 to 1.75',
                                                                          np.where((df_result['TemperatureChange'] >= 1.75) & (df_result['TemperatureChange'] < 2.0), '1.75 to 2.0',
                                                                          np.where((df_result['TemperatureChange'] >= 2.0) & (df_result['TemperatureChange'] < 2.25), '2.0 to 2.25',
                                                                          np.where((df_result['TemperatureChange'] >= 2.25) & (df_result['TemperatureChange'] < 2.5), '2.25 to 2.50',
                                                                          np.where((df_result['TemperatureChange'] >= 2.5) & (df_result['TemperatureChange'] < 2.75), '2.50 to 2.75',
                                                                          np.where((df_result['TemperatureChange'] >= 2.75) & (df_result['TemperatureChange'] < 3.0), '2.75 to 3.0',
                                                                          np.where((df_result['TemperatureChange'] >= 3.0) & (df_result['TemperatureChange'] < 3.25), '3.0 to 3.25',
                                                                          np.where((df_result['TemperatureChange'] >= 3.25) & (df_result['TemperatureChange'] < 3.5), '3.25 to 3.50',
                                                                          np.where(df_result['TemperatureChange'] >= 3.5, '>3.50',
                                                                 2))))))))))))))))))))))))

In [141]:
color_discrete_map = {
                        "<-2.0": "#023e8a",
                        "-2.0 to -1.75": "#0077b6",
                        "-1.75 to -1.50": "#0096c7",
                        "-1.50 to -1.25": "#00b4d8",
                        "-1.25 to -1.0": "#48cae4",
                        "-1.0 to -0.75": "#90e0ef",
                        "-0.75 to -0.25": "#ade8f4",
                        "-0.50 to 0.25": "#a8dadc",
                        "-0.25 to 0.0": "#caf0f8",
                        "0.0 to 0.25": "#ffd166",
                        "0.25 to 0.50": "#ffb600",
                        "0.50 to 0.75": "#ffb600",
                        "0.75 to 1.0": "#ffaa00",
                        "1.0 to 1.25": "#ff9e00",
                        "1.25 to 1.50": "#ff9100",
                        "1.50 to 1.75": "#ff8500",
                        "1.75 to 2.0": "#ff7900",
                        "2.0 to 2.25": "#ff6d00",
                        "2.25 to 2.50": "#ff6000",
                        "2.50 to 2.75": "#ff5400",
                        "2.75 to 3.0": "#ff4800",
                        "3.0 to 3.25": "#f92432",
                        "3.25 to 3.50": "#c70512",
                        ">3.50": "#9f040e",
                        }

# Libreria: plotly https://plotly.com/python/
## Referencias: https://plotly.com/python/bubble-maps/

In [152]:
import plotly.express as px
df = px.data.gapminder()
fig = px.scatter_geo(df_result, locations="Country", locationmode='country names',
                     hover_name="Country", size="Size", color="Color", color_discrete_map=color_discrete_map,
                     animation_frame="Year", hover_data='TemperatureChange', category_orders={"Color": list(color_discrete_map.keys())},
                     title="Annual Surface Temperature Change from 1961 to 2021 in Celsius Degrees",
                     projection="winkel tripel", width=1500, height=800)
fig.show()